controllers/clusterurlmonitor/clusterurlmonitor_supplement.go

```diff
@@ -47,31 +64,162 @@ func (s *ClusterUrlMonitorSupplement) EnsureFinalizer() (reconcile.Result, error
 	return reconcile.ContinueReconcile()
 }
 
-func (s *ClusterUrlMonitorSupplement) EnsureServiceMonitorExists() error {
+func (s *ClusterUrlMonitorSupplement) EnsureServiceMonitorExists() (reconcile.Result, error) {
 
 	namespacedName := types.NamespacedName{Name: s.ClusterUrlMonitor.Name, Namespace: s.ClusterUrlMonitor.Namespace}
 	exists, err := s.doesServiceMonitorExist(namespacedName)
 	if exists || err != nil {
-		return err
+		return reconcile.RequeueReconcileWith(err)
 	}
 
 	clusterDomain, err := s.getClusterDomain()
 	if err != nil {
-		return err
+		return reconcile.RequeueReconcileWith(err)
 	}
 
 	spec := s.ClusterUrlMonitor.Spec
 	clusterUrl := spec.Prefix + clusterDomain + ":" + spec.Port + spec.Suffix
 	serviceMonitor := templates.TemplateForServiceMonitorResource(clusterUrl, s.BlackBoxExporter.GetBlackBoxExporterNamespace(), namespacedName)
-	err = s.Client.Create(s.Ctx, &serviceMonitor)
+
+	// Does the resource already exist?
+	if err := s.Client.Get(s.Ctx, namespacedName, &serviceMonitor); err != nil {
+		// If this is an unknown error
+		if !k8serrors.IsNotFound(err) {
+			// return unexpectedly
+			return utilreconcile.RequeueReconcileWith(err)
+		}
+		// and create it
+		err = s.Client.Create(s.Ctx, &serviceMonitor)
+		if err != nil {
+			return utilreconcile.RequeueReconcileWith(err)
+		}
+	}
+
+	//Update status with serviceMonitorRef
+	desiredRef := v1alpha1.NamespacedName{
+		Name:      namespacedName.Name,
+		Namespace: namespacedName.Namespace,
+	}
+	emptyRef := v1alpha1.NamespacedName{}
+
+	currentRef := s.ClusterUrlMonitor.Status.ServiceMonitorRef
+	if currentRef != emptyRef && desiredRef != currentRef {
+		return utilreconcile.RequeueReconcileWith(customerrors.InvalidReferenceUpdate)
+	}
+
+	if currentRef == emptyRef && desiredRef != emptyRef {
+		s.ClusterUrlMonitor.Status.ServiceMonitorRef = desiredRef
+
+		err := s.Client.Status().Update(s.Ctx, &s.ClusterUrlMonitor)
+		if err != nil {
+			return utilreconcile.RequeueReconcileWith(err)
+		}
+		return utilreconcile.StopReconcile()
+
+	}
+	return utilreconcile.ContinueReconcile()
+}
+
+func (s *ClusterUrlMonitorSupplement) EnsurePrometheusRuleResourceExists() (utilreconcile.Result, error) {
```

Comment:
We really have quite a bit of dublicate code to RouteMonitor Controller.  I think it really makes sense at one point to add a supplement for both controllers. However, this may be too much for that ticket and surely strongly depends on how much we still want to add in terms of functionality. So you can probably ignore that comment

Reply:
I have tried to unify code from both controllers but it wasn't possible (or I didn't see the way to do it) because each of them has a set of specific required CRD fields. https://github.com/openshift/route-monitor-operator/blob/f0f94b287e9f61042ad874ee85885712a92f69f6/api/v1alpha1/routemonitor_types.go#L27